### PR TITLE
Launches a GCP instance and regist a DNS record with its public IP

### PR DESCRIPTION
**Pull request by fabioDMFerreira**
#718

* Add domain name flag to create instance command.

* Automatically creates a managed zone and the A record for the instance launched.

### CODE REVIEW

File: lepton/gcp.go [hunk 1]
```diff
@@ -420,27 +421,122 @@ func (p *GCloud) CreateInstance(ctx *Context) error {
...
+			}
+		}
+	}
+
+	// Create DNS record with instance IP
+	resource := &dns.ResourceRecordSet{
+		Name:    aRecordName,
+		Type:    "A",
+		Rrdatas: []string{aRecordIP},
+		Ttl:     21600,
```
Comment (eyberg) [inline lepton/gcp.go:532]: I'd make this 5min

File: lepton/utils.go [hunk 2]
```diff
@@ -14,3 +15,58 @@ func exitWithError(errs string) {
 func arrayToString(a []int, delim string) string {
 	return strings.Trim(strings.Replace(fmt.Sprint(a), " ", delim, -1), "[]")
 }
+
+// isDomainValid returns an error if the domain name is not valid
+// See https://tools.ietf.org/html/rfc1034#section-3.5 and
+// https://tools.ietf.org/html/rfc1123#section-2.
+// code extracted from https://gist.github.com/chmike/d4126a3247a6d9a70922fc0e8b4f4013
+func isDomainValid(name string) error {
```
Comment (eyberg) [inline lepton/utils.go:23]: this is fine to have - just curious why you ran into this?

**Review by eyberg — CHANGES_REQUESTED** [review]
i'd just adjust the ttl to 5min

File: lepton/utils.go (hunk 2 above)
Comment (fabioDMFerreira) [reply lepton/utils.go:23]: The function seemed pretty complete to grant domain names have the right format and prevent errors on creating the rules in a cloud provider.

**Comment by fabioDMFerreira** [comment]
**Heads up** support to create automatically DNS records in AWS was added to this PR.

File: lepton/aws.go [hunk 3]
```diff
@@ -521,6 +526,133 @@ func (p *AWS) CreateInstance(ctx *Context) error {
...
+		return err
+	}
+
+	dnsService := route53.New(sess)
+
+	domainParts := strings.Split(domainName, ".")
+
+	// example:
+	// domainParts := []string{"test","example","com"}
+	zoneName := domainParts[len(domainParts)-2]                 // example
```
Comment (eyberg) [inline lepton/aws.go:575]: this looks like a dupe of the gcloud one? can we make this common?

File: lepton/gcp.go [hunk 4]
```diff
@@ -423,28 +424,123 @@ func (p *GCloud) CreateInstance(ctx *Context) error {
...
+	dnsService, err := dns.NewService(context)
+	if err != nil {
+		return err
+	}
+
+	domainParts := strings.Split(domainName, ".")
+
+	// example:
+	// domainParts := []string{"test","example","com"}
+	zoneName := domainParts[len(domainParts)-2]                       // example
```
Comment (eyberg) [inline lepton/gcp.go:495]: dupe of above?

File: lepton/provider.go [hunk 5]
```diff
@@ -50,6 +58,62 @@ type VolumeService interface {
...
+	IP   string
+	Type string
+	TTL  int
+}
+
+// DNSService is an interface for DNS related operations
+type DNSService interface {
+	FindOrCreateZoneIDByName(config *Config, name string) (string, error)
+	// CreateZone(config *Config, name string) error
+	// GetZoneRecords(config *Config, zoneID string) ([]interface{}, error)
```
Comment (eyberg) [inline lepton/provider.go:73]: cruft?

**Review by eyberg — APPROVED** [review]

**Pull request merged**